### PR TITLE
prov/rxd: AV bug fixes and small message optimization

### PR DESCRIPTION
**Pull request by aingerson**
- various AV bug fixes
- eliminate the need for a CTS and data message when sending small messages (save data within RTS and ACK right away)

Verified working on ubertest test_configs/ofi_rxd/udp,test and also with data verification

### CODE REVIEW

File: prov/rxd/src/rxd.h [hunk 1]
```diff
@@ -223,6 +224,8 @@ struct rxd_ctrl_hdr {
 	uint64_t tag;
 	uint32_t op;
 	uint32_t seg_size;
+	uint8_t source[RXD_NAME_LENGTH];
+	void *inline_data;
```
Comment (shefty) [inline prov/rxd/src/rxd.h]: Don't embed pointers into protocol messages.  The value isn't usable, and the size of the pointer will differ based on the platform.

**Comment by aingerson** [comment]
Added a patch to reformat pkt structs to make more sense and to address your comment @shefty

**Comment by aingerson** [comment]
Updated to address your comments on the newest patch. For now, I've moved the version to be in the pkt_hdr so both types of packet structs have the version number. The version isn't needed for the data packets, but right now I can't think of another way to both distinguish between the two types of packets while also having the version number only in the ctrl_hdr. I'm working on a patch to decrease the size of the pkt_hdr using repeated seg_nos and this version in the data pkts can go away if we move to the cq_data path, but for now I think this is the best solution.

**Comment by shefty** [comment]
Why isn't version needed in the data packet, but is in the control packet?  How is the actual protocol version determined?

File: prov/rxd/src/rxd.h [hunk 2]
```diff
@@ -214,58 +215,66 @@ DECLARE_FREESTACK(struct rxd_x_entry, rxd_rx_fs);
...
+	uint32_t rx_id;
+	uint32_t key;
+	uint32_t seg_no;
+	fi_addr_t peer;
+};
+
+struct rxd_ctrl_hdr {
 	uint16_t type;
 	uint16_t window;
 	uint64_t size;
```
Comment (shefty) [inline prov/rxd/src/rxd.h]: This has a 32-bit gap between window and size fields.

Comment (aingerson) [reply prov/rxd/src/rxd.h]: Ahh shoot, forgot to fix it.

**Comment by aingerson** [comment]
All data packets are preceded by a control packet with the version number so it can determine the protocol when it matches the data packet to an in progress message.

**Comment by shefty** [comment]
So, can you just move the control header to the front of the control packet?  It sounds like what you need is something like this:

base_hdr
ctrl_hdr
data_hdr

With the base_hdr in front of all packets.  The ctrl packets would need all headers.  And the data packets would only need the base + data headers.

**Comment by shefty** [comment]
Typically, the very first field of any protocol packet is the version.  Otherwise, you end up severely limited in what sort of changes you can make when updating the protocol.  So, we really need to move that field first in any packet that must carry it and deal with the fallout from there.

**Comment by aingerson** [comment]
Right, that's why I changed it to how it is right now, so that the version is in the pkt_hdr along with the flags right after it. If it makes more sense, I can move the version and flags to a base_hdr before all pkts to be more like what you were proposing:
base_hdr (version, flags)
ctrl_hdr (size, data, window, etc)
data (tx_id, rx_id, seg_no, etc)
Does that make more sense?

**Comment by shefty** [comment]
Are you going to replace this PR or build from it?

**Comment by aingerson** [comment]
I'm building on it, but can easily just replace it instead. Whatever you would rather do.

**Comment by shefty** [comment]
Either way, I just need to know to merge this.

**Comment by aingerson** [comment]
Yeah, go for it. It's easier to just build off of this for me.

**Pull request merged**